enhance template schema

treywelsh opened the pull request:
Enhance template schema and share common parts with VM entity

Solve issue #15
And fix #23

jaypif commented on opennebula/resource_opennebula_template.go:
```diff
@@ -34,8 +34,24 @@ func resourceOpennebulaTemplate() *schema.Resource {
+			"graphics": graphicsSchema(),
+			"nic":      nicSchema(),
+			"os":       osSchema(),
+			"pairs": {
```
this is very close to tags (#22 ). Can you rename this as "tags" instead of pairs ?

jaypif commented on opennebula/resource_opennebula_template.go:
```diff
@@ -34,8 +34,24 @@ func resourceOpennebulaTemplate() *schema.Resource {
+			"disk":     diskSchema(),
+			"graphics": graphicsSchema(),
+			"nic":      nicSchema(),
+			"os":       osSchema(),
```
can you add:
- "hypervisors" (simple string with a validate functions) to support "KVM, vCenter or LXD"
- "SCHED_REQUIREMENTS" (as a string)
- "User inputs" section (this is to manage if cpu, vcpu or memory values are fixed or not)
- "description": to have a description part
- "vmgroup": like it is the case for VM

treywelsh replied:
Just to be sure that I don't make any mistake, all of these attributes should be shared between VM and Template schemas ?

For user input, it should be a TypeMap (we can fill USER_INPUT with custom values), as for Tags schemas. Right ?

Thank you :)

jaypif replied:
yes

jaypif commented on opennebula/resource_opennebula_template_test.go:
```diff
@@ -25,6 +25,7 @@ func TestAccTemplate(t *testing.T) {
 					resource.TestCheckResourceAttr("opennebula_template.template", "name", "terra-tpl"),
 					resource.TestCheckResourceAttr("opennebula_template.template", "permissions", "660"),
 					resource.TestCheckResourceAttr("opennebula_template.template", "group", "oneadmin"),
+					resource.TestCheckResourceAttr("opennebula_template.template", "cpu", "0.5"),
```
you should add a test function for any possible parameter (context, os, graphics, ...)

treywelsh replied:
I still need to add some tests, but anyway, for dynamic parts as context or user_inputs it may require a bit more of work are they can take custom values

jaypif commented on opennebula/shared_schemas.go:
```diff
@@ -0,0 +1,444 @@
+	}
+
+	//Generate VM Group definition
+	vmgroup := d.Get("vmgroup").(*schema.Set).List()
```
not define in schema. I think this is the reason why CI is not passing

jaypif reviewed (changes requested):
Very grat work!
However some minor changes should be done:
- can you rebase to add vmgroup ?
- Can you also update template with new parameters (specific to templates) ?

Thanks

treywelsh committed:
vm, template: switch from TypeSet to TypeList

jaypif commented on opennebula/shared_schemas.go:
```diff
@@ -0,0 +1,579 @@
+		Type:        schema.TypeList,
+		Optional:    true,
+		Computed:    true,
+		MinItems:    0,
```
useless because it is Optional ;)

jaypif commented on opennebula/shared_schemas.go:
```diff
@@ -0,0 +1,579 @@
+		Optional:    true,
+		Computed:    true,
+		ForceNew:    true,
+		MinItems:    0,
```
useless because it is Optional ;)

jaypif commented on opennebula/shared_schemas.go:
```diff
@@ -0,0 +1,579 @@
+
+	hyp, ok := d.GetOk("hypervisor")
+	if ok {
+		tpl.AddPair("HYPERVISOR", hyp.(string))
```
maybe force "ToLower" here as OpenNebula values are in lower case

jaypif reviewed (changes requested):
doc update is missing.
you change template and VM resource parameter. Your PR must have a commit with doc update.


Thanks

treywelsh committed:
update VM and Template doc and changelog

treywelsh committed:
template: tests: remove ExpectNonEmptyPlan flag

treywelsh committed:
template: share VM template building

 - deprecate template attribute
 - replace with detailed attributes common with VM
 - new schema part shared: tags
 - change os and graphics schemas to TypeList and add tests

treywelsh committed:
fix CHANGELOG and doc errors

jaypif commented on website/docs/r/template.html.markdown:
````diff
@@ -16,15 +16,67 @@ a new template will be created. When destroyed, that template will be removed.
 ```hcl
-data "template_file" "templatetpl" {
-  template = "${file("template-tpl.txt")}"
+data "template_file" "cloudinit" {
````
to be removed

jaypif commented on website/docs/r/template.html.markdown:
```diff
@@ -16,15 +16,67 @@ a new template will be created. When destroyed, that template will be removed.
+    TITLE="$TITLE",
+    NETWORK = "YES"
+    HOSTNAME = "$NAME"
+    USER_DATA = "${data.template_file.cloudinit.rendered}"
```
to be removed or replaced by a correct variable

jaypif commented on website/docs/r/template.html.markdown:
```diff
@@ -33,7 +85,7 @@ with template file `template-tpl.txt`:
 CPU = 1
```
remove this txt file as it was only needed by the deprecated parameter

jaypif commented on website/docs/r/template.html.markdown:
```diff
@@ -53,6 +105,16 @@ The following arguments are supported:
 * `template` - (Required) Text describing the OpenNebula template object, in Opennebula's XML string format.
```
now it is deprecated and optional.

jaypif commented on website/docs/r/virtual_machine.html.markdown:
```diff
@@ -144,6 +161,26 @@ The following attribute are exported:
 * `state` - State of the virtual machine.
 * `lcmstate` - LCM State of the virtual machine.
 
+## Instantiate from a template
```
This is quite confusing for this first version.

If the user instantiate a `template_id` here is the behavior:
* for all parameters excepted `context`: parameters present in VM overrides parameters defined in template
* for context: it merges them

Thanks

jaypif commented:
all goo thanks!
